### ISSUE TITLE
Query: Can skyfield (or jplephem) be used to extract (and save) a date subset of a JPL ephemeris?

### ISSUE THREAD

**Issue by Bernmeister** [skyfielders/python-skyfield#123]
I use pyephem in my Ubuntu indicator (indicator-lunar) and am trialling python-skyfield to possibly jump across in the future.

Given the ephemeris is downloaded first time by python-skyfield, or loaded from cache, it means a large download still must occur (either first time running or during the initial software install).

As my software only wants to display upcoming rise/set/position information, surely I don't need all the data in de430.bsp (something like one thousand years of data).

Does python-skyfield/python-jplephem allow me to create a subset of the .bsp data file?  Even if this is possible, would using the subset data file still be valid in python-skyfield, or is there some dependency for a "zero" date?

**Comment by brandon-rhodes** [comment]
That capability does not exist yet, but I definitely want to add it soon! Let's keep this issue open for a few weeks until I can loop back from the big refactor I'm currently doing and look at adding this capability. I'll comment here when I've made some progress!

**Comment by barrycarter** [comment]
At ftp://ssd.jpl.nasa.gov/pub//eph/planets/ascii/de430/ you can get subsets of DE430 and there's a similar URL (that I can't remember at the moment, but poke around) where you can get 1000 year subsets.

The DE430 format isn't all that complicated (it's just a bunch of Chebyshev polynomials), so you might be able to extract what you need directly.

**Comment by Bernmeister** [comment]
@barrycarter Yes, I have seen the other subsets, but they are still overkill for my needs.  At [https://naif.jpl.nasa.gov/pub/naif](https://naif.jpl.nasa.gov/pub/naif) there can be found a variety of .bsp file (and text equivalents).

There are a bunch of utilities at [https://naif.jpl.nasa.gov/naif/utilities_PC_Linux_64bit.html](https://naif.jpl.nasa.gov/naif/utilities_PC_Linux_64bit.html) but none that I've found extract subsets and re-save.

**Comment by barrycarter** [comment]
How little data do you need?

**Comment by Bernmeister** [comment]
Without sound facetious, as little as possible.

The indicator I have written displays attributes for moon, sun, planets for the current date/time and the user's position on Earth.  Data for comets/satellites are auto-downloaded to compute upcoming rise/set times.  All this is done using PyEphem.

Ultimately, I want to do the same with Skyfield.  The amount of data I would need from the .bsp file (or equivalent) would be enough to compute values for the near future.  The default de430.bsp file contains approximately 1000 years of data (500 forward and 500 back).  I don't need that much and certainly not the historical data.  So I'm looking for enough data to get me through the next decade or three, but I'm happy to do a release of the indicator software annually to update data if need be.  I just want to avoid a data file of 15 MB if possible.

**Comment by barrycarter** [comment]
Even just the data for just 1950-2050 is about 30M: http://www.filewatcher.com/b/ftp/ssd.jpl.nasa.gov/pub/eph/planets/ascii/de430-0.html but the ascp files there are fairly easy to trim (but aren't in BSP format alas).

You might look at http://space.stackexchange.com/questions/10490/is-there-a-way-to-extract-the-chebyshev-coefficients-for-a-body-from-a-spice-ker/10504 for guidance.

Or you may look at the kernels in https://naif.jpl.nasa.gov/pub/naif/generic_kernels/spk/planets/a_old_versions/ (one of which covers 2000-2020 in 2.2M) but are less accurate.

**Comment by Bernmeister** [comment]
I have made a smaller (and hopefully correct) version of the .bsp data following [these](http://space.stackexchange.com/a/10492/3442) steps (thanks to @barrycarter for the link)...

1. Download [spkmerge](https://naif.jpl.nasa.gov/naif/utilities.html) (I had to add executable permission).

2. Download [naif0010.tls](https://naif.jpl.nasa.gov/pub/naif/pds/data/co-s_j_e_v-spice-6-v1.0/cosp_1000/data/lsk/naif0010.tls).  There is another [version](https://naif.jpl.nasa.gov/pub/naif/pds/data/mex-e_m-spice-6-v1.0/mexsp_1000/DATA/LSK/NAIF0010.TLS) but I don't know if it makes a difference to the final result.

3. Download the .bsp file of your choosing from ftp://ssd.jpl.nasa.gov/pub/eph/planets/bsp.  There are a couple of readme files at ftp://ssd.jpl.nasa.gov/pub/eph/planets/bsp/README.txt and ftp://ssd.jpl.nasa.gov/pub/eph/planets/README.txt to provide background.

4. Create a text file (refer to to the [manual](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/ug/spkmerge.html) for commands) and save as `spkmergeCommand.txt`:
```
LEAPSECONDS_KERNEL = naif0010.tls
SPK_KERNEL = small.bsp
   LOG_FILE = spkmerge.log
   SOURCE_SPK_KERNEL = de421.bsp
      BEGIN_TIME = 2017-01-01
      END_TIME = 2018-01-01
```

5. From a terminal run `./spkmerge spkmergeCommand.txt` and you should have `small.bsp`.  I found that I had to delete `small.bsp` and `spkmerge.log` between runs.

I ran a test program (compute Mars' ra/dec/az/alt from a location on Earth) and got identical results using the `de421.bsp` compared with `small.bsp`.

Winner, winner, chicken dinner!

@brandon-rhodes If the steps above are correct, there is little point including this functionality within Skyfield.  Perhaps adding a section on the documentation page on how to subset data might be the quickest solution (and from my perspective, this issue can be closed).

**Comment by ckuethe** [comment]
I don't see an spkmerge binary for arm, arm64, or everything else where python could run. so including this functionality in skyfield isn't without merit.

**Comment by barrycarter** [comment]
Just to be whiny, I'll point out naif0012.tls is the latest and includes the leap second defined at the end of last year: https://naif.jpl.nasa.gov/pub/naif/generic_kernels/lsk/naif0012.tls

**Comment by brandon-rhodes** [comment]
Thanks for the link! I think I will keep this open, because the merge binary needs the whole file to be downloaded in order to do its work. My planned improvement using `jplephem` will not need the file on disk at all, but will do dynamic access to its file blocks using FTP range fetches (or HTTP range fetches) so that only the segments that the user need are pulled over the network.

**Comment by brandon-rhodes** [comment]
I have just released jplephem 2.7 which supports creating excerpts of large ephemerides. You can either invoke its command line tool on a local file:
```
python -m jplephem excerpt 2018/1/1 2018/4/1 de421.bsp outjup.bsp
```
— or you can provide a URL and jplephem will use HTTP `Range:` requests to fetch only the blocks of the ephemeris that cover the date range:
```
python -m jplephem excerpt 2018/1/1 2018/4/1 https://naif.jpl.nasa.gov/pub/naif/generic_kernels/spk/satellites/jup310.bsp  excerpt.bsp
```
Enjoy!
